### ISSUE TITLE
ReloadSchemas will remove the bitemporal axis and collections 

### ISSUE THREAD

**Issue by peetkes** [rjrudin/ml-gradle#126]
If you have your database setup to handle bitemporal documents and you have docs in the database, a schema reload will mess up your bitemporal collection as it rigorously cleans the schemas database and thus removes your bitemporal axis and collections.
After running this command I had a number of documents in a collection that was marked as bitemporal but after the ReloadSchemas, the collection was no longer marked bitemporal.
When I then redeployed the temporal configuration, which reinserts the axis and the bitemp collection in the schemas db, it complains.
Is it possible to change the ReloadSchemas so it will only remove the schemas from the database and then load the new schemas into the schemas database?

**Comment by rjrudin** [comment]
Thanks @peetkes , I'll look into this soon. Should be simple to fix - I know that reloading schemas was based on reloading modules, which also clears the modules database first. Should be easy to change that to just deleting a collection or two.

**Comment by rjrudin** [comment]
Email @grechaw for confirmation that we should just delete the http://marklogic.com/xdmp/tde collection when reloading schemas in the schemas database.

**Comment by rjrudin** [comment]
Removing this from 2.4.0 for now. I'm still not sure of the best solution with the right naming. "mlLoadSchemas" is a little misleading because it is loading schemas, but there are other things that can be loaded into the "schemas" database. "mlReloadSchemas" in theory could depend on that task and a new task for clearing just the TDE collection, but I'm not sure if that's the right approach yet.

**Comment by grechaw** [comment]
Yes, I see the point.  We clear the schemas database but don't have methods to restore things that are created in the schemas database, say by the mgmt API.

**Comment by rjrudin** [comment]
Update on this - I plan on having temporal support soon in ml-gradle (probably 2.7.0), and that will make it much easier to figure out the right solution for this.

**Comment by peetkes** [comment]
What is the status on this? 
I'm currently working on a datahub project that uses temporal protection and tde and as soon as I hit mlReloadSchemas, my temporal axis and xcollections are gone...
dhf version 3.0.0

**Comment by rjrudin** [comment]
@grechaw Do you know of a reliable way to do this? I think the issue is that a developer can have schema files that are loaded straight into the schemas database, but then there are also resources created via the Manage API that create documents in the schemas database as a side effect. 

So whereas "reload modules" is easy to support because all modules are provided by a developer, "reload schemas" seems difficult to support because it's not well-defined what resources need to be redeployed. 

An alternative is for "reload schemas" to not delete everything in the schemas database, but that seems a little misleading.

**Comment by grechaw** [comment]
Dont you in other cases crawl the filesystem and just remove what was loaded as part of the initial deploy?  I thought that security works that way.  It will definitely be more confusing than a straight cleanout, but clearing schemas I agree is much too eager.

I guess a workaround would be to have the temporal configuration in the same project -- to assert that a project has complete control over a schemas db.  Less surprising but maybe contrary to what @peetkes is trying to accomplish

**Comment by rjrudin** [comment]
Adding this to 3.9.0, I'll try to figure out the right solution here and then how to support it in ml-gradle.

**Comment by rjrudin** [comment]
@peetkes @grechaw At least for ML 9.0-x, does it seem to suffice here if "mlReloadSchemas" does the following:

1. Clears the schemas database
1. Loads the contents of src/main/ml-schemas
1. Deploys any bitemporal resources that exist
1. Deploys any view-schema resources that exist

I think bitemporal and view-schemas are the only Manage API resources that, when deployed, add documents to the schemas database? 

Are there any downsides to repeatedly deploying bitemporal stuff? Or - is it advisable to delete bitemporal documents and then recreate them? Does that have any negative impacts?

**Comment by derms** [comment]
I think it might be safer for the clear schemas command to not delete the bitemporal config (unless a flag is provided)

**Comment by peetkes** [comment]
What happens with already present bitemporal documents when you clear the schemas db and thus remove temporal config?

**Comment by miguelrgonzalez** [comment]
I think you can remove the collection, but you can not create a temporal
collection which already contains documents.

**Comment by grechaw** [comment]
I agree that the behavior is tricky to define. A set of URI patterns would enable really fine-grained control over clearing schemas.

A similar thing is happening for mlReloadModules with regard to the DHF -- there are now different sets of modules that represent user code or framework code and we should be able to manage them separately.  For 4.0 it looks like we're going to load DHF modules each time the user modules are reloaded.

**Comment by rjrudin** [comment]
@peetkes @grechaw Does it seem okay to require DMSDK (and thus ML9) for this task? I think that's the best way to delete a subset of documents, as it's guaranteed to scale. Can probably get away with using /v1/eval to execute a xdmp:document-delete or xdmp:collection-delete on a set of URIs/collections, but that seems hackier. Does have the benefit of working on ML8 though.

**Comment by rjrudin** [comment]
Collections to not delete:

- http://marklogic.com/xdmp/temporal/axis
- http://marklogic.com/xdmp/temporal/collection
- http://marklogic.com/xdmp/view (for view-schemas)
